src/jit/lower.cpp

```diff
@@ -1962,127 +2030,16 @@ void Lowering::LowerFastTailCall(GenTreeCall* call)
-
-        // Start searching in execution order list till we encounter call node
-        unsigned  tmpLclNum = BAD_VAR_NUM;
-        var_types tmpType   = TYP_UNDEF;
-        for (GenTree* treeNode = putArgStkNode->gtNext; treeNode != call; treeNode = treeNode->gtNext)
-        {
-            if (treeNode->OperIsLocal() || treeNode->OperIsLocalAddr())
-            {
-                // This should not be a GT_PHI_ARG.
-                assert(treeNode->OperGet() != GT_PHI_ARG);
-
-                GenTreeLclVarCommon* lcl = treeNode->AsLclVarCommon();
-
-                // Fast tail calling criteria permits passing of structs of size 1, 2, 4 and 8 as args.
-                // It is possible that the callerArgLclNum corresponds to such a struct whose stack slot
-                // is getting over-written by setting up of a stack arg and there are further uses of
-                // any of its fields if such a struct is type-dependently promoted.  In this case too
-                // we need to introduce a temp.
-                if ((lcl->gtLclNum == callerArgNum) || (lcl->gtLclNum == callerArgLclNum))
-                {
-                    // Create tmp and use it in place of callerArgDsc
-                    if (tmpLclNum == BAD_VAR_NUM)
-                    {
-                        // Set tmpType first before calling lvaGrabTemp, as that call invalidates callerArgDsc
-                        tmpType   = genActualType(callerArgDsc->lvaArgType());
-                        tmpLclNum = comp->lvaGrabTemp(
-                            true DEBUGARG("Fast tail call lowering is creating a new local variable"));
-
-                        comp->lvaTable[tmpLclNum].lvType            = tmpType;
-                        comp->lvaTable[tmpLclNum].lvDoNotEnregister = comp->lvaTable[lcl->gtLclNum].lvDoNotEnregister;
-                    }
-
-                    lcl->SetLclNum(tmpLclNum);
-                }
-            }
-        }
-
-        // If we have created a temp, insert an embedded assignment stmnt before
-        // the first putargStkNode i.e.
-        //     tmpLcl = CallerArg
-        if (tmpLclNum != BAD_VAR_NUM)
-        {
-            assert(tmpType != TYP_UNDEF);
-            GenTreeLclVar* local      = new (comp, GT_LCL_VAR) GenTreeLclVar(GT_LCL_VAR, tmpType, callerArgLclNum);
-            GenTree*       assignExpr = comp->gtNewTempAssign(tmpLclNum, local);
-            ContainCheckRange(local, assignExpr);
-            BlockRange().InsertBefore(firstPutArgStk, LIR::SeqTree(comp, assignExpr));
-        }
-    }
-
-    // Insert GT_START_NONGC node before the first GT_PUTARG_STK node.
-    // Note that if there are no args to be setup on stack, no need to
-    // insert GT_START_NONGC node.
     GenTree* startNonGCNode = nullptr;
-    if (firstPutArgStk != nullptr)
+    if (!putargs.Empty())
     {
+        bool     unused;
+        GenTree* insertionPoint = BlockRange().GetTreeRange(putargs.Bottom(), &unused).FirstNode();
```

Comment:
@BruceForstall Does this look OK?

Reply:
There's `jitstd::vector<T>` but that always allocates memory while `ArrayStack<T>` has an inline array that's sufficient for most purposes, especially in this code. What we really need is `Array<T>` to avoid the `Top`/`Bottom` confusion.